src/main/java/com/team1816/lib/subsystems/drive/SwerveDrive.java

```diff
@@ -1,491 +1,553 @@
+    public void resetOdometry(Pose2d pose) {
+        gyroHeading = Rotation2d.fromDegrees(infrastructure.getYaw());
+        swerveOdometry.resetPosition(gyroHeading, actualModulePositions, pose);
+        swerveOdometry.update(gyroHeading, actualModulePositions);
+        updateRobotState();
+    }
+
+    /**
+     * Zeroes the azimuth sensors of the swerve modules based on a certain pose
+     *
+     * @param pose Pose2d
+     * @see SwerveModule#zeroAzimuthSensor()
+     * @see Drive#zeroSensors()
+     */
+    @Override
+    public void zeroSensors(Pose2d pose) {
+        System.out.println("Zeroing drive sensors!");
+
+        // resetting ACTUAL module states
+        for (int i = 0; i < 4; i++) {
+            actualModuleStates[i] = new SwerveModuleState();
+            swerveModules[i].zeroAzimuthSensor();
+        }
+
+        resetOdometry(pose);
+        startingPose = pose;
+        chassisSpeed = new ChassisSpeeds();
+        isBraking = false;
+    }
+
+    /**
+     * Stops the drivetrain
+     *
+     * @see Drive#stop()
+     */
+    @Override
+    public synchronized void stop() {
+        for (int i = 0; i < 4; i++) {
+            SwerveModuleState stoppedState = new SwerveModuleState(
+                0,
+                swerveModules[i].getActualState().angle
+            );
+            swerveModules[i].setDesiredState(
+                stoppedState,
+                controlState == ControlState.OPEN_LOOP
+            );
+            desiredModuleStates[i] = stoppedState;
+        }
+    }
+
+    /** config and tests */
+
+    /**
+     * Tests the drivetrain by testing each swerve module
+     *
+     * @return true if tests passed
+     * @see SwerveModule#checkSystem()
+     * @see Drive#testSubsystem()
+     */
+    @Override
+    public boolean testSubsystem() {
+        boolean modulesPassed = true;
+        for (SwerveModule swerveModule : swerveModules) {
+            if (!swerveModule.checkSystem()) {
+                modulesPassed = false;
+                break;
+            }
+        }
+
+        return modulesPassed;
+    }
+
+    /**
+     * Returns the drive motor pid configuration of the drivetrain
+     *
+     * @return PIDSlotConfiguration
+     * @see Drive#getPIDConfig()
+     */
+    @Override
+    public PIDSlotConfiguration getPIDConfig() {
+        PIDSlotConfiguration defaultPIDConfig = new PIDSlotConfiguration();
+        defaultPIDConfig.kP = 0.0;
+        defaultPIDConfig.kI = 0.0;
+        defaultPIDConfig.kD = 0.0;
+        defaultPIDConfig.kF = 0.0;
+        return (factory.getSubsystem(NAME).implemented)
+            ? factory
+            .getSubsystem(NAME)
+            .swerveModules.drivePID.getOrDefault("slot0", defaultPIDConfig)
+            : defaultPIDConfig;
+    }
+
+    /**
+     * Returns the associated kinematics of the drivetrain
+     *
+     * @return swerveKinematics
+     */
+    public SwerveDriveKinematics getKinematics() {
+        return swerveKinematics;
```

Comment:
additionally make auto balance into a separate control mode of drive

Reply:
so setTeleOp inputs is the only thing that needs to be called in teleOp periodic with no excess logic